stable_plugins/visualization/file_types/qasm_visualization.py

```diff
@@ -0,0 +1,252 @@
+            type=PluginType.visualization,
+            entry_point=EntryPoint(
+                href=url_for(f"{QASM_BLP.name}.ProcessView"),
+                ui_href=url_for(f"{QASM_BLP.name}.MicroFrontend"),
+                plugin_dependencies=[],
+                data_input=[
+                    InputDataMetadata(
+                        data_type="*",
+                        content_type=["text/x-qasm"],
+                        parameter="data",
+                        required=True,
+                    )
+                ],
+                data_output=[
+                    DataMetadata(
+                        data_type="*",
+                        content_type=["image/svg+xml"],
+                        required=True,
+                    )
+                ],
+            ),
+            tags=[],
+        )
+
+
+@QASM_BLP.route("/ui/")
+class MicroFrontend(MethodView):
+    """Micro frontend for the QASM visualization plugin."""
+
+    @QASM_BLP.html_response(
+        HTTPStatus.OK, description="Micro frontend of the QASM visualization plugin."
+    )
+    @QASM_BLP.arguments(
+        QasmInputParametersSchema(
+            partial=True, unknown=EXCLUDE, validate_errors_as_result=True
+        ),
+        location="query",
+        required=False,
+    )
+    @QASM_BLP.require_jwt("jwt", optional=True)
+    def get(self, errors):
+        """Return the micro frontend."""
+        return self.render(request.args, errors, False)
+
+    @QASM_BLP.html_response(
+        HTTPStatus.OK, description="Micro frontend of the QASM visualization plugin."
+    )
+    @QASM_BLP.arguments(
+        QasmInputParametersSchema(
+            partial=True, unknown=EXCLUDE, validate_errors_as_result=True
+        ),
+        location="form",
+        required=False,
+    )
+    @QASM_BLP.require_jwt("jwt", optional=True)
+    def post(self, errors):
+        """Return the micro frontend with prerendered inputs."""
+        return self.render(request.form, errors, not errors)
+
+    def render(self, data: Mapping, errors: dict, valid: bool):
+        plugin = QasmVisualization.instance
+        if plugin is None:
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR)
+        schema = QasmInputParametersSchema()
+        return Response(
+            render_template(
+                "qasm_visualization.html",
+                name=plugin.name,
+                version=plugin.version,
+                schema=schema,
+                valid=valid,
+                values=data,
+                errors=errors,
+                example_values=url_for(f"{QASM_BLP.name}.MicroFrontend"),
+            )
+        )
+
+
+@QASM_BLP.route("/circuits/")
+def get_circuit_image():
+    circuit_url = request.args.get("circuit-url", None)
+    if circuit_url is None:
+        abort(HTTPStatus.NOT_FOUND)
+    filename = hashlib.md5(circuit_url.encode("utf-8")).hexdigest() + ".png"
+    path = pathlib.Path(__file__).parent.absolute() / "img" / filename
```

Comment:
Committing a `DataBlob` without explicitly setting the `id` doesn't seem to work:

```python
DataBlob(plugin_id='plugin-id', value=<binary data>)
DB.session.commit()
```

<details>
<summary>Example error</summary>

```
Traceback (most recent call last):
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask_smorest/response.py", line 89, in wrapper
    func(*args, **kwargs)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/webargs/core.py", line 594, in wrapper
    return func(*args, **kwargs)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/flask_smorest/arguments.py", line 82, in wrapper
    return func(*f_args, **f_kwargs)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/qhana_plugin_runner/api/jwt_helper.py", line 96, in wrapper
    return func(*args, **kwargs)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/stable_plugins/visualization/file_types/qasm_visualization.py", line 207, in get_circuit_image
    blob = DataBlob.set_blob(
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/qhana_plugin_runner/db/models/virtual_plugins.py", line 316, in set_blob
    DB.session.commit()
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/sqlalchemy/orm/scoping.py", line 553, in commit
    return self._proxied.commit()
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1906, in commit
    trans.commit(_to_root=True)
  File "<string>", line 2, in commit
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py", line 103, in _go
    self._raise_for_prerequisite_state(fn.__name__, current_state)
  File "/mnt/e/Temp/hiwi/qhana-plugin-runner/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 906, in _raise_for_prerequisite_state
    raise sa_exc.PendingRollbackError(
sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) NOT NULL constraint failed: DataBlob.id
[SQL: INSERT INTO "DataBlob" (plugin_id, value) VALUES (?, ?)]
[parameters: ('qasm-visualization@v0-2-0', <memory at 0x7f37ee002980>)]
(Background on this error at: https://sqlalche.me/e/20/gkpj) (Background on this error at: https://sqlalche.me/e/20/7s2a)
```
</details>

I suppose this is because `DataBlob` has a multi-column primary key:

https://github.com/UST-QuAntiL/qhana-plugin-runner/blob/24e02eaa1d36daeede9d57c61a0d5f53c8fe3f28/qhana_plugin_runner/db/models/virtual_plugins.py#L267-L273

Would it make sense to replace `DataBlob.id` (int) by `DataBlob.key` (str) such that we can reference DataBlobs by `plugin_id` and `key` pairs like for PluginStates?

Reply:
That may be because the plugin id is also part of the primary key and this prevents autoincrement from working. Yes, it is probably for the best to align this with the plugin state and change the id to a key column.